### DISCUSSION TITLE
Order of processing within series_analysis: when to apply a mask

### DISCUSSION THREAD

**Discussion by mpm-meto** [dtcenter/METplus#1925] [File I/O] [answered]
I'm running series_analysis using global inputs though I only want to focus on a sub-region of the world. I needed to use MET_CONFIG_OVERRIDES to get masking capability in the METplus wrapper as the documentation and the basic use case shows this has not been implemented in the wrapper config. The good news is it works but it does seem to insist on processing the entire grid it reads in (i.e. the global grid) and the output grid is clearly global too. See attached. It would seem beneficial to make the tool run faster to check for masks _before_ this is done and to apply the mask (like a cookie cutter) before any calculations are done. It would reduce the netcdf file size and definitely speed things up which for series_analysis would be hugely beneficial. I am only guessing here, as I haven't looked at the C++ code, but it's a matter of reorganising the order things are done in. It's a given that any regridding has to be done before calculations are done. I think the same should apply to the masking. It may be less trivial than this of course! Thoughts please. 
![s_anl_gm_all_024_ME](https://user-images.githubusercontent.com/64001904/200862938-2cfd8098-f852-4ae9-a8eb-8e925b15ae29.png)

**Comment by georgemccabe** [comment, marked as answer]
Hi @mpm-meto,
I created GitHub issue #1926 to add support for setting the mask values via the SeriesAnalysis wrapper so you don't have to use MET_CONFIG_OVERRIDES. I'll let @JohnHalleyGotway comment on how masking is handled in the MET tool itself.

**Reply by JohnHalleyGotway** [reply]
@mpm-meto I see you have a question about the processing logic of Series-Analysis. You are passing global data to it and see that the output is written to the same global domain that was used as input. And you're concerned that it may be doing extra processing that it doesn't need to do.

Here's the [line of code](https://github.com/dtcenter/MET/blob/e7f82bc26a1a641e02d3761e36fc391813b5949e/src/tools/core/series_analysis/series_analysis.cc#L758) in series_analysis.cc where the masking region is checked. This is inside the loop over the grid points in the domain. It skips any grid points for which the masking region is disabled or bad data is present. And I'm confident that that's the logic we want.

In general, *masking* in MET defines regions of interest within the verification domain. It does not *change* the verification domain. However, the **regrid** option does exactly that. Instead of using the [mask](https://github.com/dtcenter/MET/blob/e7f82bc26a1a641e02d3761e36fc391813b5949e/data/config/SeriesAnalysisConfig_default#L114) configuration option in the Series-Analysis config file, you could consider using the [regrid](https://github.com/dtcenter/MET/blob/e7f82bc26a1a641e02d3761e36fc391813b5949e/data/config/SeriesAnalysisConfig_default#L30) option. If processing data on a global lat/lon grid, defining a sub-domain as your region of interest should be pretty straight-forward. As long as the location of those input/output grid points are coincident using the NEAREST neighbor method should run quickly. The output written by Series-Analysis would be reduced in the way you'd like.

FYI, this is loosely related to a somewhat similar option in MODE called [plot_valid_flag](https://github.com/dtcenter/MET/blob/e7f82bc26a1a641e02d3761e36fc391813b5949e/data/config/MODEConfig_default#L216). But this is just a plotting option, not a regridding one. It zooms up MODE's PostScript plots to include only the region of valid data. But supplying a similar option for regridding is not so easy. For lat/lon grids, regridding to only include the region of valid data is simple enough. But doing so more generally for other projection types (lambert conformal, polar stereo, mercator, etc) is not.

If you do think an automated regridding step to "zoom" lat/lon and gaussian grids based on the masking region provided is worth pursuing, that's a GitHub issue we could add. But again, it'd be limited to only a couple of projection types and would error out for others.

**Reply by JohnHalleyGotway** [reply]
Just wanted to follow up on this discussion to see if questions remain. If not, please feel free to select the "Mark As Answer" button for one of the previous responses, and then I'll lock it to prevent future posts. If I don't hear otherwise in the next week or so, I'll just go ahead and mark it as being answered.

Thanks!
